Refactor two-way data binding using proxy

mohamedhyouns opened the pull request:
- Fix #32

mohamedhyouns committed:
refactor: two-way binding using proxy to improve input synchronization

- Refactored two-way binding logic to allow number
inputs to remain empty.

- Added `getRangePercent` function to compute slider
percentage values.

- Introduced `updateFieldValue` to handle empty or numeric
inputs gracefully.

- Implemented `sliderStateProxy` to synchronize range and
number inputs.

- Centralized input handling via `handleInputChange` to
simplify event management.

- Removed wired `addEventListener` logic to reduce bugs and
improve maintainability.

mohamedhyouns committed:
chore: generate the new assets

mohamedhyouns committed:
refactor: add 'name' attributes for proxy-based synchronization

- Added `name` attributes to slider inputs in `index.html`
for proxy-based synchronization.

mohamedhyouns committed:
fix: unit tests

mohamedhyouns committed:
chore: deprecate multi-stage event test

Removed test as the proxy now handles input and range
updates automatically.

mohamedhyouns committed:
test: add new test coverage

mohamedhyouns committed:
style: reformating the file

The current style of file not consistent with vscode reformating style

mohamedhyouns commented:
After merging this and updating the DXBR builder with the new slider files, this [bug](https://dxpr.monday.com/boards/935035417/pulses/8035873807) will be valid and need to be fixed

mohamedhyouns commented:
Resources:
- [Eirik Vullum: JavaScript Metaprogramming - ES6 Proxy Use and Abuse | JSConf Budapest 2017](https://www.youtube.com/watch?v=_5X2aB_mNp4)
- [Two-way data binding and reactivity with 15 lines of vanilla JavaScript
]( https://gomakethings.com/two-way-data-binding-and-reactivity-with-15-lines-of-vanilla-javascript/)

jjroelofs reviewed (commented):
I will also comment on the issue but the bug you found is not a bug but a feature, here is why:

1. The number input is not a real form element in this use case, it is just a tool that enhances the slider input with the possibility for more accuracy, that is why it also should not have a name attribute, we don't want to submit it separately to the server.
2. A range element's value cannot be cleared, "empty" is not a valid state for it. That is why allowing the number input to be empty would be a bug.
3. The default value of a range input is to be in the center. When clearing the number input, it slider goes to the center, and the number input assumes the value corresponding to the middle of range slider. This is expected behavior.

mohamedhyouns commented:
@jjroelofs Please have a look https://dxpr.monday.com/boards/935035417/pulses/8055708426

mohamedhyouns committed:
fix: replace `name` with `data-dxb-proxy-key`

mohamedhyouns committed:
fix: set slider in the middle if the input field is cleared

mohamedhyouns committed:
fix: set slider in the middle in both inputs when input number is empty

mohamedhyouns commented:
![image](https://github.com/user-attachments/assets/a35ae7cc-40ca-4286-bc03-d24dc74eaca7)

@jjroelofs, I got the idea of resetting both inputs to the middle when trying to clear the input of type number. This has been addressed [here](https://github.com/dxpr/DXB-Slider/pull/38/commits/402568a0b88a4c8cd66016f74b122ad9a3cffb97).

![image](https://github.com/user-attachments/assets/9532f2da-0d8c-4363-8955-e15be4ee72f9)
@jjroelofs Regarding the [current logic that fires two events](https://github.com/dxpr/DXB-Slider/blob/ad2761ea80f54a039ba4290af86105316d58add1/dxb-slider.js#L61-L69), I believe it exists as a workaround to avoid [incorrect setups of listeners in the live editing manager](https://github.com/dxpr/dxpr_builder/issues/2947). Since creating the input of type number in the DXB slider is an async process, this workaround seems to have been added to compensate. However, as a developer, I expect only one event listener to fire when interacting with either the input of type number or type range in the DXB slider. Debugging the code step by step reveals this behavior to be a bug.

![image](https://github.com/user-attachments/assets/525e4848-d6e3-4c82-aafb-e109fb128ba1)
@jjroelofs  I see the issue I’m trying to solve in the current PR is still unclear. First, let me clarify that the current issue is an atomic issue that addresses a flaw in the architecture of how the DXB slider updates values for both inputs. IMHO, we need to adopt the new logic, and it’s not feasible to split this into smaller changes due to the nature of the issue.

Since this is an atomic issue addressing multiple bugs, do you suggest releasing it as 3.x? 

I recorded a video to explain why the existing logic is problematic: https://www.youtube.com/watch?v=B8V_Bw9NK44.

jjroelofs committed:
Merge branch '2.x' into mohamedhany/2.x/#32-refactor-two-way-binding

jjroelofs reviewed (changes requested):
This PR is too complicated for me to review without having information about the engineering decisions that were made and why they were made, so I can't accept it without a good commit message and PR description. 

Also, I don't like the requirement for a new data attribute. The attribute values added in the examples are identical to the id values with already have so it seems they are redundant. Please see if you can remove this requirement because this change makes implementation more error prone and it is not backwards compatible, meaning we would have to bump the version to 3.x.

mhmelshaaerdxpr commented on dxb-slider.js:
```diff
@@ -45,34 +89,44 @@
+          return;
+        }
+
+        sliderStateProxy[e.target.dataset["dxbProxyKey"]] = e.target.value;
```
Please be consistent, always use `dataset.dxbProxyKey` for getting and setting values. Don't mix `setAttribute("data-attr-name", value)` with `dataset.attributeName`.
Also, you may put the attribute name in a const variable and use it whenever needed instead of hardcoding the same attribute name in all uses.

mohamedhyouns replied:
https://github.com/dxpr/DXB-Slider/pull/38/commits/5aa8e3350bcf46d58604c3b767771be86ac91083
https://github.com/dxpr/DXB-Slider/pull/38/commits/68ee689c5cbceaac0e48a8d262e0ff390790e1e8

mhmelshaaerdxpr commented on dxb-slider.js:
```diff
@@ -45,34 +89,44 @@
-        updateValue();
-        rangeInput.dispatchEvent(new Event('input', { bubbles: true }));
-      });
+      [rangeInput, numberInput].forEach(input =>
+        input.addEventListener('input', handleInputChange)
+      );
```
Not sure if it's a good idea to have the same handler. For instance, we need to separate the `numberInput` handler and introduce a debounce to invoke the handler only after the user stops writing values—i.e., a `500ms` debounce.

mohamedhyouns replied:
https://github.com/dxpr/DXB-Slider/pull/38/commits/a4a1d4c6cfec7a26b084b7a4f27494475c0864b7
Adding debounce won't be a good idea, it will affect the UX I tested it with 500 and without denouncing 

https://github.com/user-attachments/assets/4d2afacf-1661-4c05-880f-6751cb00c908

mhmelshaaerdxpr reviewed (changes requested):
The new changes introduce new bugs:
1. _Investment Amount_ slider example is broken.

    ![image](https://github.com/user-attachments/assets/10f837e1-1aec-48e3-8f6d-8f16dffa0ff5)
2. The input validation is no longer working—i.e., you can set values outside the configured range or fraction of step value. 

    ![image](https://github.com/user-attachments/assets/47c5acb3-bd4d-4d5f-a876-5f9cfb25242d)
    ![image](https://github.com/user-attachments/assets/f2a8639a-172c-494c-8e30-ac6ccb4ab329)

mohamedhyouns committed:
fix: prevent input value from exceeding maximum value

mohamedhyouns committed:
fix: support new slider `investment`

mhmelshaaerdxpr commented on dxb-slider.js:
```diff
@@ -97,6 +97,14 @@
           return;
         }
 
+        // Reset the input value to the previously stored value if it exceeds the maximum allowed value
+        const newValue = Number(e.target.value);
+        const max = e.target.max;
+
+        if (max && newValue > e.target.max) {
+          e.target.value = sliderStateProxy[e.target.dataset["dxbProxyKey"]]
+        }
+
```
Please hand this over to the proxy not the event listener handler. This when Javascript Proxy are actually used for—i.e., validation.

mohamedhyouns replied:
https://github.com/dxpr/DXB-Slider/pull/38/commits/a4496c08e53b84eaa400a907ed3a5b437925b199

mohamedhyouns committed:
fix: deprecate `data-dxb-proxy-key` data attribute

mohamedhyouns committed:
fix: `MutationObserver` infinite loop

mohamedhyouns committed:
fix: move reset max logic to proxy

mohamedhyouns committed:
chore: decouple input listeners

mohamedhyouns committed:
chore: update dist file

mhmelshaaerdxpr reviewed (changes requested):
Please test your code thoroughly. This is the second instance where easily detectable bugs could have been caught with basic testing.

- For ranges like [-100, 100] with default step, I cannot add negative values (it resets to mid range).
- Value resets are confusing: Sometimes reset to max (i.e., when exceed maximum), reset to mid (i.e., on invalid input or on clear).
- In the 100 step example, putting a fraction of a step value do approximation on the range field while keep the number as is in the input field. Did we decide at some point to do this approximation part? Regardless, there is a mismatch in behavior between the input and range fields.
- We should provide feedback rather than do corrective actions on validation errors. This is the standard on how input fields behaves.

Although you introduce some code enhancements, I have the feeling that the new logic is a bit messy and results in more bugs. I will have to double check your final implementation after you fix the bugs.

mohamedhyouns committed:
feat: improve input validation and handling in `dxb-slider.js`

- UPDATED: Enhanced min/max value validation to handle empty
attributes correctly
- UPDATED: Converted `validInput.min` and `validInput.max` to numbers
for consistent comparisons
- ADDED: Allow temporary entry of negative (`"-"`) and floating point
 (`"."`) characters before validation
- FIXED: Prevent premature value correction while the user is
still typing

mohamedhyouns committed:
chore: update tests

mohamedhyouns committed:
fix: allow empty/clear input

mohamedhyouns committed:
chore: update dist file

mohamedhyouns commented:
> * For ranges like [-100, 100] with default step, I cannot add negative values (it resets to mid range).

The slider supports negative value: https://github.com/dxpr/DXB-Slider/pull/38/commits/7a415141f00e9737fcab5fe29cc26ffdc780939d

> * Value resets are confusing: Sometimes reset to max (i.e., when exceed maximum), reset to mid (i.e., on invalid input or on clear).

I agree, but not in all mentioned cases.

-  reset to max (i.e., when exceed maximum) -> This is a MUST since the input is bound by the slider. Please take a look at this  [component](https://formkit.com/inputs/slider#linked-number-inputs).
- Clear now is allowed: https://github.com/dxpr/DXB-Slider/pull/38/commits/4e88e93082453274ac2f6dca4378b0a7b9c4845d and https://github.com/dxpr/DXB-Slider/pull/38/commits/09d663a4b6734e6dd32d322bc7a13123c759df9b

> * In the 100 step example, putting a fraction of a step value do approximation on the range field while keep the number as is in the input field. Did we decide at some point to do this approximation part? Regardless, there is a mismatch in behavior between the input and range fields.

~~Sorry but didn't understand your point.~~

I asked ChatGPT about the proper Rounding Formula, and it gave me this 🙂
![image](https://github.com/user-attachments/assets/24a3ab6e-7200-4140-8619-1ea875278a34)

![image](https://github.com/user-attachments/assets/09c89890-bf02-46b3-8977-5c8286320ebe)

Seems to work fine for me: https://github.com/dxpr/DXB-Slider/pull/38/commits/1354a0cb7399f81c941724b4afdd70446843f878



> * We should provide feedback rather than do corrective actions on validation errors. This is the standard on how input fields behaves.

The core implementation doesn't support this at the moment. Adding it would require a new issue, as it involves some complexity and is difficult to incorporate into the current atomic PR.

mohamedhyouns committed:
chore: remove `getDefaultValue`

mohamedhyouns committed:
fix: allow emptiness when clearing in case of min isn't Zero

mohamedhyouns committed:
refactor: improve value handling and precision in `updateFieldValue`

- ADDED: `roundToPrecision` function to maintain correct
floating-point precision
- UPDATED: Ensured number input allows empty values without
applying calculations
- UPDATED: Added explicit parsing for `min`, `max`, and `step`
attributes to prevent unintended behavior
- UPDATED: Clamped input values within the allowed range before
applying rounding
- UPDATED: Improved rounding logic to correctly align with
step increments
- FIXED: Ensured precision consistency when updating number input values

mohamedhyouns committed:
Merge branch '2.x' into mohamedhany/2.x/#32-refactor-two-way-binding

mohamedhyouns commented:
Some references:
- https://formkit.com/inputs/slider#linked-number-inputs
- https://observablehq.com/@observablehq/synchronized-inputs
